### PR TITLE
DI Disposing and Project Format Upgrade

### LINKED ISSUE #2112: Document whether injected objects should be disposed
[The DI document](http://docs.asp.net/en/latest/fundamentals/dependency-injection.html#service-lifetimes-and-registration-options) should make it very clear whether we need to dispose injected objects, or whether we can rely on the built-in IoC container to do that for us.

I've looked into the source code, so I know what is happening behind the scenes, but I cannot rely on that behavior.

Please document this.

### PR DESCRIPTION
**Pull request by ardalis**
Fixes #2112 

* Updated to VS2017 project format
* Added description of how types are disposed

I didn't update the actual code samples to include cover disposability; I just added an inline code section.

Perhaps @davidfowl could review.

### CODE REVIEW

File: aspnetcore/fundamentals/dependency-injection.md [hunk 1]
```diff
@@ -156,7 +156,7 @@ Scoped lifetime services are created once per request.
...
 
-Services can be registered with the container in several ways. We have already seen how to register a service implementation with a given type by specifying the concrete type to use. In addition, a factory can be specified, which will then be used to create the instance on demand. The third approach is to directly specify the instance of the type to use, in which case the container will never attempt to create an instance.
+Services can be registered with the container in several ways. We have already seen how to register a service implementation with a given type by specifying the concrete type to use. In addition, a factory can be specified, which will then be used to create the instance on demand. The third approach is to directly specify the instance of the type to use, in which case the container will never attempt to create an instance (nor will it dispose of the instance).
```
Comment (davidfowl) [inline aspnetcore/fundamentals/dependency-injection.md]: 👍

File: aspnetcore/fundamentals/dependency-injection.md [hunk 2]
```diff
@@ -211,6 +211,28 @@ What if you find that your classes tend to have way too many dependencies being
...
 With regards to data access specifically, you can inject the `DbContext` into your controllers (assuming you've added EF to the services container in `ConfigureServices`). Some developers prefer to use a repository interface to the database rather than injecting the `DbContext` directly. Using an interface to encapsulate the data access logic in one place can minimize how many places you will have to change when your database changes.
 
+### Disposing of services
+
+The container will call `Dispose` for `IDisposable` types it creates. However, if you add an instance to the container yourself, it will not be disposed.
+
+Example:
+
+```csharp
+// Assume Service1, Service2, and Service3 each implement IDisposable
```
Comment (davidfowl) [inline aspnetcore/fundamentals/dependency-injection.md]: Should we show the class definition?

Comment (ardalis) [inline aspnetcore/fundamentals/dependency-injection.md]: It just seemed like clutter. I can add if you want, but it doesn't impact the sample any (except to show what implementing IDisposable looks like, but I think customers who care about this behavior will already be familiar with that). Your call.

Comment (Rick-Anderson) [inline aspnetcore/fundamentals/dependency-injection.md]: @ardalis  yes, show the class defn.

Comment (ardalis) [inline aspnetcore/fundamentals/dependency-injection.md]: It shall be done.

**Comment by ardalis** [comment]
@Rick-Anderson this should be good to merge now.

**Pull request merged**